### PR TITLE
Multiple DLE backends and rudimentary Yask support

### PR DESCRIPTION
**Pull request by FabioLuporini**
Basic infrastructure to support multiple DLE backends. Currently we have:
- the usual DLE
- a rudimentary YASK backend

Nothing essentially works on the YASK side apart from a few Devito AST nodes that can successfully be converted to YASK AST nodes, so there are no tests yet. The next PR will finalize this AST to AST conversion and also provide the relevant tests.

You see lots of edits but it's mostly just reshuffling code and making a better use of the OOP (eg hierarchy of DLE backends)

### CODE REVIEW

**Comment by FabioLuporini** [comment]
Rebased on master.

If there are no major objections, I'd like to merge this @mlange05 @navjotk

File: devito/dle/backends/advanced.py [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+                mapper[candidate] = List(body=fissioned)
+
+            processed.append(Transformer(mapper).visit(node))
+
+        return {'nodes': processed}
+
+    @dle_pass
+    def _loop_blocking(self, state, **kwargs):
+        """
+        Apply loop blocking to :class:`Iteartion` trees.
```
Comment (mlange05) [inline devito/dle/backends/advanced.py]: Typo, should be `Iteration`

File: devito/dle/backends/advanced.py [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+        self._create_elemental_functions(state)
+        self._loop_blocking(state)
+        self._simdize(state)
+        if self.params.get('openmp') is True:
+            self._ompize(state)
+
+    @dle_pass
+    def _loop_fission(self, state, **kwargs):
+        """
+        Apply loop fission to innermost :class:`Iteartion` objects. This pass
```
Comment (mlange05) [inline devito/dle/backends/advanced.py]: Typo, should be `Iteration`

File: devito/dle/transformer.py [hunk 3]
```diff
@@ -1,82 +1,55 @@
...
+                     * 'basic': Add instructions to avoid denormal numbers and create
+                                elemental functions for rapid JIT-compilation.
+                     * 'advanced': 'basic', vectorization, loop blocking.
+                     * '3D-advanced': Like 'advanced', but apply 3D loop blocking
+                                      if there are at least the perfectly nested
+                                      parallel iteration spaces.
+                     * 'speculative': Apply all of the 'advanced' transformations,
+                                      plus other transformations that might increase
+                                      (or possibly decrease) performance.
+                     * 'yask': Optimize by offloading to the YASK optimizer.
```
Comment (mlange05) [inline devito/dle/transformer.py]: Maybe a warning that this is not complete yet?

File: devito/dle/transformer.py [hunk 4]
```diff
@@ -1,82 +1,55 @@
...
-                    * 'blockshape': A tuple representing the shape of a block created
-                                    by loop blocking.
-                    * 'blockinner': By default, loop blocking is not applied to the
-                                    innermost dimension of an Iteration/Expression tree
-                                    to maximize the chances of SIMD vectorization. To
-                                    force the blocking of this loop, the ``blockinner``
-                                    flag should be set to True.
-                    * 'openmp': True to emit OpenMP code, False otherwise.
+    :param mode: Drive the tree transformation. ``mode`` is a string indicating
+                 a certain optimization pipeline. The following values are accepted: ::
```
Comment (mlange05) [inline devito/dle/transformer.py]: So I take creating custom optimisation sequence is no longer available, and users have to rely on the preset sequences?

Comment (FabioLuporini) [inline devito/dle/transformer.py]: Oh no, it's still possible. In fact, it's even simpler: look at ``class DevitoCustomRewriter``. Although it's still missing some bindings, the infrastructure to write arbitrary optimisation sequences is already there, and it's also being tested by three different tests

**Comment by FabioLuporini** [comment]
Will apply the fixes later today, and then merge this once Travis is green.

**Comment by FabioLuporini** [comment]
Integrated your fixes. Also rebased on master.

@navjotk any final words before landing?

File: devito/dle/transformer.py (hunk 4 above)
Comment (mlange05) [inline devito/dle/transformer.py]: Cool, thanks.

File: devito/dle/backends/advanced.py [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+                handle = functions[0]
+                dim = handle.indices[-1]
+                size = handle.shape[-1]
+                if any(dim != i.indices[-1] for i in functions):
+                    # Dangerous for correctness
+                    continue
+
+                stencils = promote_scalar_expressions(stencils, (size,), (dim,), True)
+
+                assert len(stencils) == len(expressions)
```
Comment (navjotk) [inline devito/dle/backends/advanced.py]: Asserts are ignored in python's optimised mode. Plus you want to inform the application code that something went wrong - which might not always be the same as stopping execution.

File: devito/dle/backends/advanced.py [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+                key = lambda i: i.is_Parallel and not i.is_Vectorizable
+                candidates = filter_iterations(tree, key=key, stop='consecutive')
+                if not candidates:
+                    continue
+
+                # Heuristic: if at least two parallel loops are available and the
+                # physical core count is greater than self.thresholds['collapse'],
+                # then omp-collapse the loops
+                nparallel = len(candidates)
+                if psutil.cpu_count(logical=False) < self.thresholds['collapse'] or\
```
Comment (navjotk) [inline devito/dle/backends/advanced.py]: The thresholds could possibly be added to `parameters`?

File: devito/dle/backends/common.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+
+    """
+    Bag of thresholds, to be used to trigger or prevent certain transformations.
+    """
+    thresholds = {
+        'collapse': 32,  # Available physical cores
+        'elemental': 30,  # Operations
+        'max_fission': 80,  # Statements
+        'min_fission': 1  # Statements
+    }
```
Comment (navjotk) [inline devito/dle/backends/common.py]: These thresholds could be moved to `parameters` for configurability.

File: devito/dle/transformer.py [hunk 8]
```diff
@@ -1,82 +1,56 @@
...
+                                        by loop blocking.
+                        * 'blockinner': By default, loop blocking is not applied to the
+                                        innermost dimension of an Iteration/Expression
+                                        tree to maximize vectorization. Set this flag to
+                                        True to override this heuristic.
+                        * 'openmp': True to emit OpenMP code, False otherwise.
     :param compiler: Compiler class used to perform JIT compilation. Useful to
                      introduce compiler-specific vectorization pragmas.
     """
+    assert mode is None or isinstance(mode, str)
```
Comment (navjotk) [inline devito/dle/transformer.py]: Same problem with `assert`s as above. Should be an exception in my opinion.

File: devito/dle/backends/advanced.py (hunk 5 above)
Comment (FabioLuporini) [inline devito/dle/backends/advanced.py]: I prefer to use ``assert``s when it comes to catch developer bugs. I agree, however, that stopping execution at that point would be bad. This is not part of this PR though (that piece of code was just moved); taking note, will fix in another round

File: devito/dle/backends/advanced.py (hunk 6 above)
Comment (FabioLuporini) [inline devito/dle/backends/advanced.py]: yes, I suggest we have a key-value pair in ``parameters`` such as ``(dle: {...}``, where the ``{...}`` is a dictionary with parameters concerning the DLE. Again, this PR just moves code around, so taking note for subsequent improvement

File: devito/dle/backends/common.py (hunk 7 above)
Comment (FabioLuporini) [inline devito/dle/backends/common.py]: yes (see above)

File: devito/dle/transformer.py (hunk 8 above)
Comment (FabioLuporini) [inline devito/dle/transformer.py]: this one should really be an exception, I agree. I'm gonna fix it now, because an assert is really bad ad that point. The first one you pointed out, however, is conceptually more an assert than an exception, I think (anyway, just nitpicking on my side)

**Comment by FabioLuporini** [comment]
> I left some comments. Nothing major.

Thanks

**Comment by FabioLuporini** [comment]
Addressed your comments; Travis' happy. So merging...

**Comment by FabioLuporini** [comment]
Merged.
